media-proxy/include/mesh/CONCEPT.md

```diff
@@ -0,0 +1,271 @@
+# Concept – New Architecture of Media Proxy to Support Multipoint Groups
+
+## Connection (base class)
+* Interface for Transmitting data – _may be implemented in the derived class_.
+* Interface for Receiving data – _may be implemented in the derived class_.
+* State – _When changes, triggers an Event_.
+  * `Not Configured` – _After initialization._
+  * `Configured`
+  * `Establishing`
+  * `Active`
+  * `Suspended`
+  * `Closing`
+  * `Closed`
+  * `Deleting`
+* Operations
+  * `Create` – _Performed by constructor_.
+  * `Configure`
+  * `Establish`
+  * `Suspend` – _For test/debug purposes_.
+  * `Resume` – _For test/debug purposes_.
+  * `Shutdown`
+  * `Delete` – _Initiated externally. Can't be performed by the Connection class_.
+* Status – _When changes, triggers an Event_.
+  * `Initial`
+  * `Transition`
+  * `Healthy`
+  * `Failure`
+  * `Shutdown`
+* Base metrics
+  * Throughput – _Move calculation to the Telemetry extractor level_.
+  * Transaction counter.
+  * Byte counter.
+  * Error counter.
+  * Status.
+  * State.
+  * Maciej suggested:
+    * Backend load / capacity metric. Report example: `load: 85%` – _Secondary_.
+    * Interface to enable/disable metrics by name.
+  * Tomasz suggested:
+    * Unified logger interface.
+* Information
+  * Created at.
+  * Established at.
+* Interface for sending metrics.
+* Metrics Reset interface – _Access from MCM Agent, option to clear all or certain metrics_.
+
+### Intermediary Layer: ST2110, RDMA, Local
+* Base functionality of the specific backend.
+
+### Backend Layer: (xxx)-Tx, (xxx)-Rx
+
+* Transmitter or Receiver functionality of the specific backend.
+
+## Connection Lifecycle State Diagram
+
+```mermaid
+stateDiagram
+    notinit: Not Initialized
+    configuring: Configuring
+    configured: Configured
+    establishing: Establishing
```

Comment:
do we really need those transition states like: Configuring, Establishing, Deleting? Can't we go directly to configured, established etc. ?

Reply:
Let's take a look at the `Configure` flow. The base class initiates a transition by setting the state to `Configuring`. Then, the derived class is responsible for the actual action to apply a configuration. According to the result of that action, the derived class sets the state to `Configured` or `Not Initialized`.

Now the `Establish` flow. The base class initiates a transition by setting the state to `Establishing`. Then, the derived class is responsible for the actual action to establish a connection. According to the result of that action, the derived class sets the state to `Active` or `Closed`.

The mentioned actions in the derived class can consume time, few moments or seconds. During that period, the current state will represent the fact that the action is being taken. So, yes, we need those transition states.

Reply:
fair enough